File: btrader/packages/engine-core/src/calc.ts
// ============================================================================
//  Pure trading math — no I/O, fully unit-testable. This is the core of the
//  "MT5-like" account model: margin, P/L, equity, margin level, stop-out.
// ============================================================================

import { OrderSide } from '@btrader/shared';

export interface SymbolCalcSpec {
  digits: number;
  pipSize: number; // e.g. 0.0001
  contractSize: number; // units per 1.0 lot, e.g. 100000
  marginRate: number; // multiplier (1 = full notional / leverage)
  // #14: crypto/metals margin fraction of notional, independent of account
  // leverage. When set (> 0), margin = notional * marginPercent. null = FX
  // leverage-based margin.
  marginPercent?: number | null;
  quoteCurrency: string;
  baseCurrency: string;
}

/** Smallest price increment (a "point") = 10^-digits. */
export function pointSize(digits: number): number {
  return Math.pow(10, -digits);
}

/**
 * Notional value of a position in the QUOTE currency.
 *   notional = volume(lots) * contractSize * price
 */
export function notionalQuote(volumeLots: number, spec: SymbolCalcSpec, price: number): number {
  return volumeLots * spec.contractSize * price;
}

/**
 * Required margin in the ACCOUNT currency.
 *   margin = (volume * contractSize * openPrice * marginRate) / leverage * fx(quote→acct)
 * `quoteToAcct` converts one unit of quote currency into the account currency
 * (1 when quote === account currency).
 */
export function requiredMargin(
  volumeLots: number,
  spec: SymbolCalcSpec,
  price: number,
  leverage: number,
  quoteToAcct = 1,
): number {
  const rawNotional = notionalQuote(volumeLots, spec, price);
  // #14: crypto & metals hold a fixed fraction of notional as margin, independent
  // of the account's currency leverage (and of marginRate). e.g. marginPercent
  // 0.10 → 10% → 10x effective, whatever the account leverage is.
  if (spec.marginPercent != null && spec.marginPercent > 0) {
    return rawNotional * spec.marginPercent * quoteToAcct;
  }
  if (leverage <= 0) leverage = 1;
  const notional = rawNotional * spec.marginRate;
  return (notional / leverage) * quoteToAcct;
}

/**
 * Floating / realized profit in the ACCOUNT currency for a position.
 *   BUY:  (close - open) * volume * contractSize
 *   SELL: (open - close) * volume * contractSize
 * Then convert quote→account.
 */
export function positionProfit(
  side: OrderSide,
  volumeLots: number,
  openPrice: number,
  currentPrice: number,
  spec: SymbolCalcSpec,
  quoteToAcct = 1,
): number {
  const direction = side === 'BUY' ? 1 : -1;
  const priceDiff = (currentPrice - openPrice) * direction;
  const profitQuote = priceDiff * volumeLots * spec.contractSize;
  return profitQuote * quoteToAcct;
}

/** Value of one pip for a given volume, in account currency. */
export function pipValue(volumeLots: number, spec: SymbolCalcSpec, quoteToAcct = 1): number {
  return spec.pipSize * volumeLots * spec.contractSize * quoteToAcct;
}

// ── Trading-group pricing: spread markup + dealing commission ───────────────

export type GroupCommissionType = 'NONE' | 'PER_LOT' | 'PER_SIDE' | 'ROUND_TURN' | 'PERCENT';
export type SymbolPricingMethod = 'SPREAD_ONLY' | 'COMMISSION_ONLY' | 'SPREAD_AND_COMMISSION';

export interface GroupPricing {
  markupPoints: number; // extra points applied to the traded side (broker edge)
  slippagePoints: number; // execution slippage penalty (points) worsening every fill
  commissionType: GroupCommissionType;
  commissionValue: number;
  /** Spread Markup floor (points) from a Symbol Mapping (0 = unused). */
  minSpreadPoints?: number;
  /** Spread Markup ceiling (points); 0 = fixed at min (no LP-driven growth). */
  maxSpreadPoints?: number;
  pricingMethod?: SymbolPricingMethod;
  /** #2B: group execution model — 'MARKET' (fill at market) or 'INSTANT' (honour clicked price / requote). Undefined ⇒ MARKET. */
  executionMode?: 'MARKET' | 'INSTANT';
  /** #2B: max points the market may move from the client's quote before an INSTANT order requotes; 0 ⇒ fall back to the symbol's slippagePoints. */
  instantDeviationPoints?: number;
}

export interface SymbolMappingPricing {
  pricingMethod: SymbolPricingMethod;
  minSpreadPoints: number;
  maxSpreadPoints: number;
  commissionType: GroupCommissionType;
  commissionValue: number;
}

/**
 * Live Spread Markup (points) per Symbol Group Markup Engine Spec §5.1.
 *
 * Min / Max are **markup** amounts (not a total-spread band):
 * - Normal markets → applied markup ≈ Minimum
 * - When LP spread widens → markup may rise with it, never above Maximum
 * - COMMISSION_ONLY → 0 (raw feed spread; commission charged separately)
 *
 * Returns the **per-side** points the engine applies to the fill price.
 * Quote UI applies the same value to bid (−) and ask (+) so total spread
 * widens by `2 * perSide` when both sides move — for a configured markup M,
 * perSide = M / 2 so the visible total extra equals M (spec point shift).
 *
 * When min=max=0 under a mapping, markup is 0 (no silent legacy fallback).
 * `legacyMarkupPoints` is only used when there is no mapping band at all
 * (caller passes it for pre-mapping groups).
 */
export function spreadMarkupFromBand(
  lpSpreadPoints: number,
  method: SymbolPricingMethod | undefined,
  minSpreadPoints: number,
  maxSpreadPoints: number,
  legacyMarkupPoints = 0,
): number {
  if (method === 'COMMISSION_ONLY') return 0;
  const min = Math.max(0, minSpreadPoints || 0);
  const max = Math.max(0, maxSpreadPoints || 0);
  if (min === 0 && max === 0) {
    return method === 'SPREAD_ONLY' || method === 'SPREAD_AND_COMMISSION' || !method
      ? Math.max(0, legacyMarkupPoints)
      : 0;
  }
  // Spec: floor at min; track LP spread upward; cap at max (max=0 → fixed min).
  const lp = Math.max(0, lpSpreadPoints);
  let totalMarkup = min;
  if (max > 0) {
    totalMarkup = Math.min(max, Math.max(min, lp));
  }
  // Per-side so bid/ask each move by half → total extra spread ≈ totalMarkup.
  return totalMarkup / 2;
}

/** Resolve commission from a mapping's pricing method (NONE when spread-only). */
export function commissionFromMapping(
  mapping: SymbolMappingPricing | null | undefined,
  fallback: { commissionType: GroupCommissionType; commissionValue: number },
): { commissionType: GroupCommissionType; commissionValue: number } {
  if (!mapping) return fallback;
  if (mapping.pricingMethod === 'SPREAD_ONLY') {
    return { commissionType: 'NONE', commissionValue: 0 };
  }
  return {
    commissionType: mapping.commissionType ?? 'NONE',
    commissionValue: mapping.commissionType === 'NONE' ? 0 : Number(mapping.commissionValue || 0),
  };
}

/**
 * Apply a group's spread markup to the raw LP fill price. The markup always
 * moves the price against the client (broker edge): a BUY fills higher, a SELL
 * fills lower. `points` are price points (10^-digits).
 */
export function applyMarkup(
  side: OrderSide,
  rawPrice: number,
  markupPoints: number,
  digits: number,
): number {
  if (!markupPoints) return rawPrice;
  const delta = markupPoints * pointSize(digits);
  return side === 'BUY' ? rawPrice + delta : rawPrice - delta;
}

/**
 * Dealing commission in the ACCOUNT currency, returned as a NEGATIVE number to
 * store on the position (it flows into equity immediately and realizes at
 * close). Charged in full at open:
 *   PER_LOT  : value × lots               (round-turn)
 *   PER_SIDE : value × lots × 2           (open + close booked up front)
 *   PERCENT  : value% × notional × 2      (both sides)
 */
export function dealingCommission(
  pricing: GroupPricing,
  volumeLots: number,
  spec: SymbolCalcSpec,
  openPrice: number,
  quoteToAcct = 1,
): number {
  const v = pricing.commissionValue || 0;
  switch (pricing.commissionType) {
    case 'PER_LOT':
    case 'ROUND_TURN':
      return -(v * volumeLots);
    case 'PER_SIDE':
      return -(v * volumeLots * 2);
    case 'PERCENT': {
      const notionalAcct = notionalQuote(volumeLots, spec, openPrice) * quoteToAcct;
      return -((v / 100) * notionalAcct * 2);
    }
    default:
      return 0;
  }
}

/**
 * Split a position's accrued swap and commission across a partial close.
 *
 * Swap and commission sit on the position and are realized into balance in
 * proportion to the lots being closed. Whatever is realized MUST be removed
 * from the position, or the remainder keeps carrying charges the client has
 * already paid — every subsequent close then bills them again. Four equal
 * partial closes of an un-split position charge the harmonic series
 * (1 + 1/2 + 1/3 + 1/4) ≈ 2.08x the true commission.
 *
 * `remaining*` is computed by subtraction rather than by its own ratio, so the
 * two halves sum back to the original to within one float ULP — exact once
 * stored at the column's Decimal(28,8) — instead of each side rounding
 * independently and leaving a residue that grows with every partial close.
 *
 * @param swap        accrued swap on the position (signed)
 * @param commission  commission on the position (negative = charged)
 * @param closeVolume lots being closed now
 * @param openVolume  lots on the position before this close
 */
export function splitPositionCharges(
  swap: number,
  commission: number,
  closeVolume: number,
  openVolume: number,
): {
  realizedSwap: number;
  realizedCommission: number;
  remainingSwap: number;
  remainingCommission: number;
} {
  // A full (or over-) close realizes everything; guard against divide-by-zero.
  if (!(openVolume > 0) || closeVolume >= openVolume) {
    return {
      realizedSwap: swap,
      realizedCommission: commission,
      remainingSwap: 0,
      remainingCommission: 0,
    };
  }
  const ratio = closeVolume / openVolume;
  const realizedSwap = swap * ratio;
  const realizedCommission = commission * ratio;
  return {
    realizedSwap,
    realizedCommission,
    remainingSwap: swap - realizedSwap,
    remainingCommission: commission - realizedCommission,
  };
}

export interface OpenPositionView {
  /** Position id — required to act on a view (e.g. pick a stop-out victim). */
  id?: string;
  side: OrderSide;
  volume: number;
  openPrice: number;
  currentPrice: number;
  marginUsed: number;
  swap: number;
  commission: number;
  spec: SymbolCalcSpec;
  quoteToAcct?: number;
}

export interface AccountAggregates {
  floatingPL: number;
  margin: number;
  equity: number;
  freeMargin: number;
  marginLevel: number; // %, 0 when no margin used
}

/**
 * Resolve the factor converting one unit of `quoteCcy` into `accountCcy`, or
 * null when the feed carries no usable conversion pair.
 *
 * Returning null rather than 1 is the whole point: a silent fallback to 1 books
 * margin and P/L in the wrong currency (a USD account trading EURGBP with no
 * GBPUSD quote was out by ~27%) and the error compounds for as long as the
 * position is open. Callers decide what to do — refuse when opening new risk,
 * warn and carry on when valuing something already open.
 *
 * @param mid returns the mid price of a symbol, or null if not quoted
 */
export function resolveFxFactor(
  accountCcy: string,
  quoteCcy: string,
  mid: (symbol: string) => number | null,
): number | null {
  if (!quoteCcy || !accountCcy || quoteCcy === accountCcy) return 1;
  const direct = mid(`${quoteCcy}${accountCcy}`); // e.g. GBP -> USD via GBPUSD
  if (direct != null && direct > 0) return direct;
  const inverse = mid(`${accountCcy}${quoteCcy}`); // e.g. JPY -> USD via USDJPY
  if (inverse != null && inverse > 0) return 1 / inverse;
  return null;
}

/**
 * Live P/L of one open position in the account currency, inclusive of the swap
 * and commission carried on it — the same quantity that flows into equity.
 */
export function positionFloating(p: OpenPositionView): number {
  return (
    positionProfit(p.side, p.volume, p.openPrice, p.currentPrice, p.spec, p.quoteToAcct ?? 1) +
    p.swap +
    p.commission
  );
}

/**
 * The open position carrying the largest loss — the stop-out victim.
 *
 * Selection MUST be done on live floating P/L. `Position.profit` in the
 * database is only written when a position closes, so ordering open rows by
 * that column sorts an all-zero field and liquidates an arbitrary position,
 * potentially a profitable one.
 *
 * Ties resolve to the earliest position in the input, so callers that read
 * positions in a stable order (openedAt) get deterministic, FIFO behaviour.
 *
 * @param skipIds positions to exclude — used to pass over one that failed to
 *                close so a single unclosable position cannot stall the cascade.
 */
export function worstPosition(
  positions: OpenPositionView[],
  skipIds?: ReadonlySet<string>,
): OpenPositionView | null {
  let worst: OpenPositionView | null = null;
  let worstPL = Infinity;
  for (const p of positions) {
    if (skipIds && p.id && skipIds.has(p.id)) continue;
    const pl = positionFloating(p);
    if (pl < worstPL) {
      worstPL = pl;
      worst = p;
    }
  }
  return worst;
}

/**
 * Recompute live account aggregates from balance+credit and the set of open
 * positions. Called by the engine on every relevant tick.
 *   equity      = balance + credit + Σ floatingPL + Σ swap + Σ commission
 *   margin      = Σ marginUsed
 *   freeMargin  = equity - margin
 *   marginLevel = margin > 0 ? equity / margin * 100 : 0
 */
export function computeAggregates(
  balance: number,
  credit: number,
  positions: OpenPositionView[],
): AccountAggregates {
  let floatingPL = 0;
  let margin = 0;
  for (const p of positions) {
    floatingPL += positionFloating(p);
    margin += p.marginUsed;
  }
  const equity = balance + credit + floatingPL;
  const freeMargin = equity - margin;
  const marginLevel = margin > 0 ? (equity / margin) * 100 : 0;
  return { floatingPL, margin, equity, freeMargin, marginLevel };
}

/**
 * Cash an account may withdraw, given its balance and CURRENT free margin.
 *
 *   available = max(0, min(balance, freeMargin))
 *
 * Bounded by balance because credit (bonus) is not withdrawable, and by free
 * margin because funds backing an open position are not the client's to take.
 *
 * The caller must pass a `freeMargin` derived from live prices. Deriving it
 * from `Position.profit` treats open positions as flat and lets a client
 * withdraw an unrealized loss — the balance then goes negative when the
 * position closes.
 */
export function withdrawableCash(balance: number, freeMargin: number): number {
  return Math.max(0, Math.min(balance, freeMargin));
}

/** Round a volume to the symbol lot step and clamp to [min,max]. */
export function normalizeVolume(
  volume: number,
  minLot: number,
  maxLot: number,
  lotStep: number,
): number {
  const steps = Math.round(volume / lotStep);
  let v = steps * lotStep;
  v = Math.max(minLot, Math.min(maxLot, v));
  // fix floating error
  return Math.round(v / lotStep) * lotStep;
}

/**
 * Snap volume to lot step for NEW opens. Unlike normalizeVolume, never upgrades
 * undersized lots to minLot (that would open more risk than the client asked).
 * Returns null when volume is non-positive, off-step beyond float noise, or
 * outside [minLot, maxLot].
 */
export function snapVolume(
  volume: number,
  minLot: number,
  maxLot: number,
  lotStep: number,
): number | null {
  if (!(volume > 0) || !Number.isFinite(volume) || !(lotStep > 0)) return null;
  if (!(minLot > 0) || maxLot < minLot) return null;
  const steps = Math.round(volume / lotStep);
  let v = steps * lotStep;
  v = Math.round(v / lotStep) * lotStep;
  // Reject if snap drifted too far from the request (client sent junk / wrong step).
  if (Math.abs(volume - v) > lotStep * 0.51 + 1e-9) return null;
  if (v + 1e-12 < minLot || v - 1e-12 > maxLot) return null;
  return v;
}

/** Round a price to symbol digits. */
export function roundPrice(price: number, digits: number): number {
  const f = Math.pow(10, digits);
  return Math.round(price * f) / f;
}

/**
 * Apply slippage tolerance for a market/one-click order. Returns the worst
 * acceptable fill price; engine rejects if market is beyond it.
 *   slippagePoints = 0  → no slippage allowed (default per spec).
 */
export function slippageBound(
  side: OrderSide,
  requestedPrice: number,
  slippagePoints: number,
  digits: number,
): number {
  const slip = slippagePoints * pointSize(digits);
  // BUY fills at ask; tolerate higher. SELL fills at bid; tolerate lower.
  return side === 'BUY' ? requestedPrice + slip : requestedPrice - slip;
}

File: btrader/apps/trading-core-rs/src/calc.rs
//! Pure trading math — same formulas as Node `packages/engine-core/src/calc.ts`.

#[derive(Debug, Clone)]
pub struct SymbolCalcSpec {
    pub digits: i32,
    pub pip_size: f64,
    pub contract_size: f64,
    pub margin_rate: f64,
    pub margin_percent: Option<f64>,
    pub quote_currency: String,
    pub base_currency: String,
}

pub fn point_size(digits: i32) -> f64 {
    10f64.powi(-digits)
}

pub fn notional_quote(volume_lots: f64, spec: &SymbolCalcSpec, price: f64) -> f64 {
    volume_lots * spec.contract_size * price
}

pub fn required_margin(
    volume_lots: f64,
    spec: &SymbolCalcSpec,
    price: f64,
    mut leverage: f64,
    quote_to_acct: f64,
) -> f64 {
    let raw = notional_quote(volume_lots, spec, price);
    if let Some(pct) = spec.margin_percent {
        if pct > 0.0 {
            return raw * pct * quote_to_acct;
        }
    }
    if leverage <= 0.0 {
        leverage = 1.0;
    }
    (raw * spec.margin_rate / leverage) * quote_to_acct
}

pub fn position_profit(
    side: &str,
    volume_lots: f64,
    open_price: f64,
    current_price: f64,
    spec: &SymbolCalcSpec,
    quote_to_acct: f64,
) -> f64 {
    let direction = if side.eq_ignore_ascii_case("BUY") { 1.0 } else { -1.0 };
    (current_price - open_price) * direction * volume_lots * spec.contract_size * quote_to_acct
}

#[derive(Debug, Clone, Default)]
pub struct GroupPricing {
    pub markup_points: f64,
    pub slippage_points: f64,
    pub commission_type: String,
    pub commission_value: f64,
    pub min_spread_points: f64,
    pub max_spread_points: f64,
    pub pricing_method: Option<String>,
    pub execution_mode: String,
    pub instant_deviation_points: f64,
    pub execution_delay_ms: i32,
    pub execution_apply_to: serde_json::Value,
}

/// Whether Instant honour / Market delay applies to `kind`.
/// Empty `{}` ⇒ all kinds (legacy). Otherwise only explicit `true` keys apply;
/// missing or `false` ⇒ immediate market fill (no delay / no Instant honour).
/// Admin always saves the full checkbox map; selective checks must not leak delay
/// onto unchecked SL/TP/pending kinds.
pub fn execution_applies(flags: &serde_json::Value, kind: &str) -> bool {
    let Some(obj) = flags.as_object() else {
        return true;
    };
    if obj.is_empty() {
        return true;
    }
    match obj.get(kind) {
        Some(v) => v.as_bool() == Some(true),
        None => false,
    }
}

pub fn pending_type_to_apply_kind(order_type: &str, side: &str) -> &'static str {
    let t = order_type.to_ascii_uppercase();
    let s = side.to_ascii_uppercase();
    match t.as_str() {
        "BUY_LIMIT" => "buyLimit",
        "SELL_LIMIT" => "sellLimit",
        "BUY_STOP" => "buyStop",
        "SELL_STOP" => "sellStop",
        "LIMIT" => {
            if s == "BUY" {
                "buyLimit"
            } else {
                "sellLimit"
            }
        }
        "STOP" | "STOP_LIMIT" => {
            if s == "BUY" {
                "buyStop"
            } else {
                "sellStop"
            }
        }
        _ => "buyLimit",
    }
}

pub fn market_execution_delay_ms(pricing: &GroupPricing, kind: &str) -> i32 {
    if pricing.execution_mode.to_ascii_uppercase() != "MARKET" {
        return 0;
    }
    // SL/TP never wait under MARKET delay. Waiting left the position OPEN while
    // the chart printed through the stop; fill already honours the level when
    // apply-to includes sl/tp. Group ms still applies to market + pending + close.
    if kind.eq_ignore_ascii_case("sl") || kind.eq_ignore_ascii_case("tp") {
        return 0;
    }
    if !execution_applies(&pricing.execution_apply_to, kind) {
        return 0;
    }
    pricing.execution_delay_ms.max(0)
}

pub fn spread_markup_from_band(
    lp_spread_points: f64,
    method: Option<&str>,
    min_spread_points: f64,
    max_spread_points: f64,
    legacy_markup_points: f64,
) -> f64 {
    if method == Some("COMMISSION_ONLY") {
        return 0.0;
    }
    let min = min_spread_points.max(0.0);
    let max = max_spread_points.max(0.0);
    if min == 0.0 && max == 0.0 {
        return if method == Some("SPREAD_ONLY")
            || method == Some("SPREAD_AND_COMMISSION")
            || method.is_none()
        {
            legacy_markup_points.max(0.0)
        } else {
            0.0
        };
    }
    let lp = lp_spread_points.max(0.0);
    let mut total = min;
    if max > 0.0 {
        total = max.min(min.max(lp));
    }
    total / 2.0
}

pub fn apply_markup(side: &str, raw_price: f64, markup_points: f64, digits: i32) -> f64 {
    if markup_points == 0.0 {
        return raw_price;
    }
    let delta = markup_points * point_size(digits);
    if side.eq_ignore_ascii_case("BUY") {
        raw_price + delta
    } else {
        raw_price - delta
    }
}

pub fn dealing_commission(
    pricing: &GroupPricing,
    volume_lots: f64,
    spec: &SymbolCalcSpec,
    open_price: f64,
    quote_to_acct: f64,
) -> f64 {
    let v = pricing.commission_value;
    match pricing.commission_type.to_ascii_uppercase().as_str() {
        "PER_LOT" | "ROUND_TURN" => -(v * volume_lots),
        "PER_SIDE" => -(v * volume_lots * 2.0),
        "PERCENT" => {
            let notional = notional_quote(volume_lots, spec, open_price) * quote_to_acct;
            -((v / 100.0) * notional * 2.0)
        }
        _ => 0.0,
    }
}

pub fn split_position_charges(
    swap: f64,
    commission: f64,
    close_volume: f64,
    open_volume: f64,
) -> (f64, f64, f64, f64) {
    if !(open_volume > 0.0) || close_volume >= open_volume {
        return (swap, commission, 0.0, 0.0);
    }
    let ratio = close_volume / open_volume;
    let realized_swap = swap * ratio;
    let realized_commission = commission * ratio;
    (
        realized_swap,
        realized_commission,
        swap - realized_swap,
        commission - realized_commission,
    )
}

#[derive(Debug, Clone)]
pub struct OpenPositionView {
    pub id: String,
    pub side: String,
    pub volume: f64,
    pub open_price: f64,
    pub current_price: f64,
    pub margin_used: f64,
    pub swap: f64,
    pub commission: f64,
    pub spec: SymbolCalcSpec,
    pub quote_to_acct: f64,
}

pub fn position_floating(p: &OpenPositionView) -> f64 {
    position_profit(
        &p.side,
        p.volume,
        p.open_price,
        p.current_price,
        &p.spec,
        p.quote_to_acct,
    ) + p.swap
        + p.commission
}

pub fn worst_position<'a>(
    positions: &'a [OpenPositionView],
    skip: &std::collections::HashSet<String>,
) -> Option<&'a OpenPositionView> {
    let mut worst: Option<&OpenPositionView> = None;
    let mut worst_pl = f64::INFINITY;
    for p in positions {
        if skip.contains(&p.id) {
            continue;
        }
        let pl = position_floating(p);
        if pl < worst_pl {
            worst_pl = pl;
            worst = Some(p);
        }
    }
    worst
}

#[derive(Debug, Clone, Default)]
pub struct AccountAggregates {
    pub floating_pl: f64,
    pub margin: f64,
    pub equity: f64,
    pub free_margin: f64,
    pub margin_level: f64,
}

pub fn compute_aggregates(balance: f64, credit: f64, positions: &[OpenPositionView]) -> AccountAggregates {
    let mut floating_pl = 0.0;
    let mut margin = 0.0;
    for p in positions {
        floating_pl += position_floating(p);
        margin += p.margin_used;
    }
    let equity = balance + credit + floating_pl;
    let free_margin = equity - margin;
    let margin_level = if margin > 0.0 { equity / margin * 100.0 } else { 0.0 };
    AccountAggregates {
        floating_pl,
        margin,
        equity,
        free_margin,
        margin_level,
    }
}

pub fn resolve_fx_factor(
    account_ccy: &str,
    quote_ccy: &str,
    mut mid: impl FnMut(&str) -> Option<f64>,
) -> Option<f64> {
    if quote_ccy.is_empty() || account_ccy.is_empty() || quote_ccy == account_ccy {
        return Some(1.0);
    }
    if let Some(direct) = mid(&format!("{quote_ccy}{account_ccy}")) {
        if direct > 0.0 {
            return Some(direct);
        }
    }
    if let Some(inverse) = mid(&format!("{account_ccy}{quote_ccy}")) {
        if inverse > 0.0 {
            return Some(1.0 / inverse);
        }
    }
    None
}

pub fn snap_volume(volume: f64, min_lot: f64, max_lot: f64, lot_step: f64) -> Option<f64> {
    if !(volume > 0.0) || !volume.is_finite() || !(lot_step > 0.0) {
        return None;
    }
    if !(min_lot > 0.0) || max_lot < min_lot {
        return None;
    }
    let steps = (volume / lot_step).round();
    let mut v = steps * lot_step;
    v = (v / lot_step).round() * lot_step;
    if (volume - v).abs() > lot_step * 0.51 + 1e-9 {
        return None;
    }
    if v + 1e-12 < min_lot || v - 1e-12 > max_lot {
        return None;
    }
    Some(v)
}

pub fn normalize_volume(volume: f64, min_lot: f64, max_lot: f64, lot_step: f64) -> f64 {
    let steps = (volume / lot_step).round();
    let mut v = steps * lot_step;
    v = v.max(min_lot).min(max_lot);
    (v / lot_step).round() * lot_step
}

pub fn round_price(price: f64, digits: i32) -> f64 {
    let f = 10f64.powi(digits);
    (price * f).round() / f
}

pub fn slippage_bound(side: &str, requested: f64, slippage_points: f64, digits: i32) -> f64 {
    let slip = slippage_points * point_size(digits);
    if side.eq_ignore_ascii_case("BUY") {
        requested + slip
    } else {
        requested - slip
    }
}

pub fn round_lots(v: f64, step: f64) -> f64 {
    let s = if step > 0.0 { step } else { 0.01 };
    ((v / s).round() * s * 10000.0).round() / 10000.0
}

#[cfg(test)]
mod tests {
    use super::*;

    fn eurusd() -> SymbolCalcSpec {
        SymbolCalcSpec {
            digits: 5,
            pip_size: 0.0001,
            contract_size: 100_000.0,
            margin_rate: 1.0,
            margin_percent: None,
            quote_currency: "USD".into(),
            base_currency: "EUR".into(),
        }
    }

    #[test]
    fn margin_1_lot() {
        let m = required_margin(1.0, &eurusd(), 1.1, 100.0, 1.0);
        assert!((m - 1100.0).abs() < 1e-6);
    }

    #[test]
    fn profit_buy() {
        let p = position_profit("BUY", 1.0, 1.1, 1.105, &eurusd(), 1.0);
        assert!((p - 500.0).abs() < 1e-6);
    }

    #[test]
    fn profit_sell() {
        let p = position_profit("SELL", 1.0, 1.1, 1.105, &eurusd(), 1.0);
        assert!((p + 500.0).abs() < 1e-6);
    }

    #[test]
    fn delay_instant_is_zero() {
        let p = GroupPricing {
            execution_mode: "INSTANT".into(),
            execution_delay_ms: 200,
            execution_apply_to: serde_json::json!({"marketBuy": true, "sl": true}),
            ..Default::default()
        };
        assert_eq!(market_execution_delay_ms(&p, "marketBuy"), 0);
        assert_eq!(market_execution_delay_ms(&p, "sl"), 0);
    }

    #[test]
    fn delay_only_checked_apply_to() {
        let p = GroupPricing {
            execution_mode: "MARKET".into(),
            execution_delay_ms: 150,
            execution_apply_to: serde_json::json!({
                "marketBuy": true,
                "marketSell": false,
                "sl": false,
                "tp": false,
                "buyLimit": false,
                "sellLimit": false,
                "buyStop": false,
                "sellStop": false,
                "manualClose": false,
                "closeAll": false
            }),
            ..Default::default()
        };
        assert_eq!(market_execution_delay_ms(&p, "marketBuy"), 150);
        assert_eq!(market_execution_delay_ms(&p, "marketSell"), 0);
        assert_eq!(market_execution_delay_ms(&p, "sl"), 0);
        assert_eq!(market_execution_delay_ms(&p, "tp"), 0);
        assert_eq!(market_execution_delay_ms(&p, "buyStop"), 0);
    }

    #[test]
    fn delay_empty_apply_to_means_all() {
        let p = GroupPricing {
            execution_mode: "MARKET".into(),
            execution_delay_ms: 80,
            execution_apply_to: serde_json::json!({}),
            ..Default::default()
        };
        assert_eq!(market_execution_delay_ms(&p, "sl"), 0);
        assert_eq!(market_execution_delay_ms(&p, "tp"), 0);
        assert_eq!(market_execution_delay_ms(&p, "marketBuy"), 80);
        assert_eq!(market_execution_delay_ms(&p, "buyStop"), 80);
    }

    #[test]
    fn protective_never_takes_market_delay() {
        let p = GroupPricing {
            execution_mode: "MARKET".into(),
            execution_delay_ms: 100,
            execution_apply_to: serde_json::json!({"sl": true, "tp": true, "marketBuy": true}),
            ..Default::default()
        };
        assert_eq!(market_execution_delay_ms(&p, "sl"), 0);
        assert_eq!(market_execution_delay_ms(&p, "tp"), 0);
        assert_eq!(market_execution_delay_ms(&p, "marketBuy"), 100);
    }
}
